### PR TITLE
[online-dpo] allow parse-args as list of floats

### LINKED ISSUE #2106: xpo can not work
### System Info

When I used the parameters below for training, the following error occurred. How can I modify it?
My parameters:
export CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7
deepspeed --num_gpus 8 --master_port=29501  examples/scripts/nash_md.py \
    --config_file examples/accelerate_configs/deepspeed_zero2.yaml \
    --model_name_or_path  /workspace/qwen2.5-14B-sft \
    --reward_model_path /workspace/Qwen2.5-14B-Instruct \
    --dataset_name /workspace/ultrafeedback-prompt.json \
    --learning_rate 5.0e-7 \
    --beta 0.1 \
    --output_dir /workspace/qwen2.5-14B-xpo \
    --per_device_train_batch_size 2 \
    --gradient_accumulation_steps 4 \
    --num_train_epochs 3 \
    --max_new_tokens 2048 \
    --warmup_ratio 0.1 \
    --missing_eos_penalty 1.0 \
    --overwrite_output_dir \
    --logging_steps 10 \
    --save_steps 50 \
    --bf16 \
    --ddp_timeout 180000000

The error message is:
Traceback (most recent call last):
  File "/workspace/trl/examples/scripts/nash_md.py", line 67, in <module>
    parser = TrlParser((DPOScriptArguments, NashMDConfig, ModelConfig))
  File "/workspace/trl/trl/commands/cli_utils.py", line 217, in __init__
    super().__init__(parsers)
  File "/workspace/transformers/src/transformers/hf_argparser.py", line 137, in __init__
    self._add_dataclass_arguments(dtype)
  File "/workspace/transformers/src/transformers/hf_argparser.py", line 264, in _add_dataclass_arguments
    self._parse_dataclass_field(parser, field)
  File "/workspace/transformers/src/transformers/hf_argparser.py", line 160, in _parse_dataclass_field
    raise ValueError(
ValueError: Only `Union[X, NoneType]` (i.e., `Optional[X]`) is allowed for `Union` because the argument parser only supports one type per argument. Problem encountered in field 'beta'.

Below are some other questions. Please provide guidance：
1. Does XPO support LoRA and 4-bit quantization? How can models like 14B or 72B use XPO?
2. Does XPO support DeepSpeed?

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder
- [ ] My own task or dataset (give details below)

### Reproduction

no

### Expected behavior

no

### PR DESCRIPTION
**Pull request by kashif**
# What does this PR do?

fixes #2106

### CODE REVIEW

**Comment by qgallouedec** [comment]
What about this instead?

```python
from transformers import HfArgumentParser, TrainingArguments
from dataclasses import dataclass, field
from typing import List

@dataclass
class MyArgs(TrainingArguments):
    beta: List[float] = field(default_factory=lambda: [0.1])

    def __post_init__(self):
        super().__post_init__()
        if len(self.beta) == 1:
            self.beta = self.beta[0]

parser = HfArgumentParser(MyArgs)
args = parser.parse_args_into_dataclasses()[0]
print(args.beta)
```


```
$ python 2106.py --output_dir tmp --beta 0.1
0.1
$ python 2106.py --output_dir tmp --beta 0.1 0.2
[0.1, 0.2]
$ python 2106.py --output_dir tmp
0.1
```

**Comment by kashif** [comment]
ah yes good idea! fixing

**Comment by qgallouedec** [comment]
To allow using args without the parser, we should check if the object has is sizeable as well:

```python
from transformers import TrainingArguments
from dataclasses import dataclass, field

from typing import List


@dataclass
class MyArgs(TrainingArguments):
    beta: List[float] = field(default_factory=lambda: [0.1])

    def __post_init__(self):
        super().__post_init__()
        if hasattr(self.beta, "__len__") and len(self.beta) == 1:
            self.beta = self.beta[0]


args = MyArgs(output_dir="tmp", beta=0.2)
print(args.beta) # 0.2
args = MyArgs(output_dir="tmp", beta=[0.1, 0.2])
print(args.beta) # [0.1, 0.2]
```

File: trl/trainer/nash_md_config.py [hunk 1]
```diff
@@ -26,10 +26,14 @@ class NashMDConfig(OnlineDPOConfig):
     Subclass of [`OnlineDPOConfig`] we can use all its arguments and add the following:
 
     Parameters:
-        mixture_coef (`float` or `list[float]`, *optional*, defaults to `0.5`):
-            Logit mixture coefficient for the model and reference model. If a list of floats is provided then the
-            mixture coefficient is selected for each new epoch and the last coefficient is used for the rest of the
-            epochs.
+        mixture_coef (`List[float]`, defaults to `[0.5]`):
+            Logit mixture coefficient for the model and reference model. If a list of floats is provided then the mixture
+            coefficient is selected for each new epoch and the last coefficient is used for the rest of the epochs.
```
Comment (qgallouedec) [inline trl/trainer/nash_md_config.py:31]:
```suggestion
            Logit mixture coefficient for the model and reference model. If a list of floats is provided then the
            mixture coefficient is selected for each new epoch and the last coefficient is used for the rest of the epochs.
```

File: trl/trainer/online_dpo_config.py [hunk 2]
```diff
@@ -41,7 +41,7 @@ class OnlineDPOConfig(TrainingArguments):
             Penalty applied to the score when the model fails to generate an EOS token. This is useful to encourage
             to generate completions shorter than the maximum length (`max_new_tokens`). The penalty must be a positive
             value.
-        beta (`float` or `list[float]`, *optional*, defaults to `0.1`):
+        beta (`List[float]`, *optional*, defaults to `[0.1]`):
```
Comment (qgallouedec) [inline trl/trainer/online_dpo_config.py:44]:
```suggestion
        beta (`float` or `List[float]`, *optional*, defaults to `0.1`):
```

File: trl/trainer/nash_md_config.py [hunk 3]
```diff
@@ -26,10 +26,14 @@ class NashMDConfig(OnlineDPOConfig):
     Subclass of [`OnlineDPOConfig`] we can use all its arguments and add the following:
 
     Parameters:
-        mixture_coef (`float` or `list[float]`, *optional*, defaults to `0.5`):
-            Logit mixture coefficient for the model and reference model. If a list of floats is provided then the
-            mixture coefficient is selected for each new epoch and the last coefficient is used for the rest of the
-            epochs.
+        mixture_coef (`List[float]`, defaults to `[0.5]`):
```
Comment (qgallouedec) [inline trl/trainer/nash_md_config.py:29]:
```suggestion
        mixture_coef (`float` or `List[float]`, defaults to `0.5`):
```

File: trl/trainer/nash_md_config.py [hunk 4]
```diff
@@ -26,10 +26,15 @@ class NashMDConfig(OnlineDPOConfig):
     Subclass of [`OnlineDPOConfig`] we can use all its arguments and add the following:
 
     Parameters:
-        mixture_coef (`float` or `list[float]`, *optional*, defaults to `0.5`):
+        mixture_coef (`float` or `List[float]`, defaults to `0.5`):
```
Comment (qgallouedec) [inline trl/trainer/nash_md_config.py:29]:
```suggestion
        mixture_coef (`float` or `List[float]`, *optional*, defaults to `0.5`):
```

File: trl/trainer/xpo_config.py [hunk 5]
```diff
@@ -26,8 +26,13 @@ class XPOConfig(OnlineDPOConfig):
     Subclass of [`OnlineDPOConfig`] we can use all its arguments and add the following:
 
     Parameters:
-        alpha (`float` or `List[float]`, *optional*, defaults to `1e-5`):
+        alpha (`float` or `List[float]`, defaults to `1e-5`):
```
Comment (qgallouedec) [inline trl/trainer/xpo_config.py:29]:
```suggestion
        alpha (`float` or `List[float]`, *optional*, defaults to `1e-5`):
```

File: trl/trainer/online_dpo_config.py [hunk 6]
```diff
@@ -41,7 +41,7 @@ class OnlineDPOConfig(TrainingArguments):
             Penalty applied to the score when the model fails to generate an EOS token. This is useful to encourage
             to generate completions shorter than the maximum length (`max_new_tokens`). The penalty must be a positive
             value.
-        beta (`float` or `list[float]`, *optional*, defaults to `0.1`):
+        beta (`float` or `list[float]`, *optional*, defaults to `[0.1]`):
```
Comment (qgallouedec) [inline trl/trainer/online_dpo_config.py:44]:
```suggestion
        beta (`float` or `list[float]`, *optional*, defaults to `0.1`):
```

File: trl/trainer/xpo_config.py [hunk 7]
```diff
@@ -26,8 +26,13 @@ class XPOConfig(OnlineDPOConfig):
     Subclass of [`OnlineDPOConfig`] we can use all its arguments and add the following:
 
     Parameters:
-        alpha (`float` or `List[float]`, *optional*, defaults to `1e-5`):
+        alpha (`float` or `List[float]`, *optional*, defaults to `[1e-5]`):
```
Comment (qgallouedec) [inline trl/trainer/xpo_config.py:29]:
```suggestion
        alpha (`float` or `List[float]`, *optional*, defaults to `1e-5`):
```

**Comment by qgallouedec** [comment]
I'll just add a few tests

**Review by qgallouedec — APPROVED** [review]

**Comment by qgallouedec** [comment]
LGTM thanks, just make sure to update the title

**Pull request merged**